### PR TITLE
Job plugin type ext1

### PR DESCRIPTION
**Pull request by gschueler**
Enhancements for job plugin type

**Is this a bugfix, or an enhancement? Please describe.**

* [x] add execution reference to event, project name to execution reference
* [x] allow modifying context in beforeJobStarts
* [x] add result status info for afterJobEnds event

todo:

* [ ] merge contexts inside the JobPluginService if there are multiple plugins...

### CODE REVIEW

File: rundeckapp/grails-app/domain/rundeck/Execution.groovy [hunk 1]
```diff
@@ -407,5 +410,30 @@ class Execution extends ExecutionContext {
         }
         exec
     }
+
+    ExecutionReference asReference(Closure<String> genTargetNodes = null) {
+        JobReferenceImpl jobRef = null
+        if (scheduledExecution) {
+            jobRef = new JobReferenceImpl(
+                    id: scheduledExecution.extid,
+                    scheduledExecutionName: scheduledExecution.jobName,
```
Comment (ronaveva) [inline rundeckapp/grails-app/domain/rundeck/Execution.groovy]: There is no "scheduledExecutionName" property in JobReferenceImpl
```suggestion
                    jobName: scheduledExecution.jobName,
```

File: core/src/main/java/com/dtolabs/rundeck/plugins/jobs/JobExecutionEventImpl.java [hunk 2]
```diff
@@ -3,36 +3,49 @@
...
 
     public JobExecutionEventImpl(StepExecutionContext executionContext) {
         if(executionContext != null){
             this.executionContext = ExecutionContextImpl.builder(executionContext).build();
         }
     }
 
-    public JobExecutionEventImpl() {
+    public JobExecutionEventImpl(StepExecutionContext executionContext, ExecutionReference execution) {
+        this.executionContext = executionContext;
```
Comment (ronaveva) [inline core/src/main/java/com/dtolabs/rundeck/plugins/jobs/JobExecutionEventImpl.java]: I'd prefer not to use the reference directly. Send a new one instead and the merge with the original.
```suggestion
        this.executionContext = ExecutionContextImpl.builder(executionContext).build();
```

Comment (gschueler) [reply core/src/main/java/com/dtolabs/rundeck/plugins/jobs/JobExecutionEventImpl.java]: ok, won't hurt to do it that way

File: core/src/main/java/com/dtolabs/rundeck/core/execution/WorkflowExecutionServiceThread.java [hunk 3]
```diff
@@ -83,15 +87,20 @@ public void run() {
 
     public WorkflowExecutionResult runWorkflow() {
         try {
-            JobExecutionEventImpl beforeEvent = new JobExecutionEventImpl(context);
-            iJobPluginService.beforeJobStarts(beforeEvent);
+            JobEventStatus jobEventStatus = iJobPluginService.beforeJobStarts(new JobExecutionEventImpl(context,executionReference));
+            StepExecutionContext executionContext = jobEventStatus.getExecutionContext();
```
Comment (gschueler) [inline core/src/main/java/com/dtolabs/rundeck/core/execution/WorkflowExecutionServiceThread.java]:
```suggestion
            StepExecutionContext executionContext = jobEventStatus != null ? jobEventStatus.getExecutionContext() : null;
```

**Pull request merged**